src/Consumption/LimitAttemptsExtension.php

```diff
@@ -0,0 +1,80 @@
+<?php
+declare(strict_types=1);
+
+namespace Cake\Queue\Consumption;
+
+use Cake\Queue\Job\Message;
+use Enqueue\Consumption\Context\MessageResult;
+use Enqueue\Consumption\MessageResultExtensionInterface;
+use Enqueue\Consumption\Result;
+use Interop\Queue\Processor;
+
+class LimitAttemptsExtension implements MessageResultExtensionInterface
+{
+    /**
+     * The property key used to set the number of times a message was attempted.
+     *
+     * @var string
+     */
+    public const ATTEMPTS_PROPERTY = 'attempts';
+
+    /**
+     * The maximum number of times a job may be attempted. $maxAttempts defined on a
+     * Job will override this value.
+     *
+     * @var int|null
+     */
+    protected $maxAttempts;
+
+    /**
+     * @param int|null $maxAttempts The maximum number of times a job may be attempted.
+     * @return void
+     */
+    public function __construct(?int $maxAttempts = null)
+    {
+        $this->maxAttempts = $maxAttempts;
+    }
+
+    /**
+     * @param \Enqueue\Consumption\Context\MessageResult $context The result of the message after it was processed.
+     * @return void
+     */
+    public function onResult(MessageResult $context): void
+    {
+        if ($context->getResult() !== Processor::REQUEUE) {
+            return;
+        }
+
+        $message = $context->getMessage();
+
+        $jobMessage = new Message($message, $context->getContext());
+
+        $maxAttempts = $jobMessage->getMaxAttempts() ?? $this->maxAttempts;
```

Comment:
Couldn't we return early on line 43 if `$this->maxAttempts === null`?

Reply:
The way I wrote this was to allow `maxAttempts` defined on the job to override `maxAttempts` defined on the extension. So if `$this->maxAttempts` is null, as long as `$jobMessage->getMaxAttempts()` is not null, there will be a maxAttempts value to work with.